### PR TITLE
Refactor statistics

### LINKED ISSUE #3256: refactor statistics
Stat gathering (fails, passes, errors, duration) is too closely coupled to the reporter.  The reporter shouldn't be the single source of truth for this information, yet it is.  Statistics should be *queried by the reporters* instead of just dumped into the base reporter.  The result may be as simple as an abstraction between the runner and reporter.

This makes it difficult to modify the stats we gather, or how we gather them.   

related: #3223, #3250

### PR DESCRIPTION
**Pull request by craigtaub**
### Description of the Change

- a very small abstraction of stats collecting into a dedicated module
- "runner" now initiates collecting
- presentation is still handled by reporter
- no context is maintained internally, so invoked as normal function

### Alternate Designs

- an API for querying could be introduced but felt overkill unless others thought it worthwhile
- currently called from Base reporter but if it is useful the reporters could call it directly themselves (so we could interchange collectors).
- I was not sure if this should be called by Base reporter or the runner, so I chose reporter. If anyone else has opinions please let me know

### Benefits

- stats handling is not so closely tied to the Base reporter,
- easier to extend, test, isolate, enhance stats collecting

### Applicable issues

Fixes #3256
(semver-minor)

### CODE REVIEW

**Comment by craigtaub** [comment]
Sorry about failing test, will investigate.

**Comment by plroebuck** [comment]
Basic question: why is missing runner argument (to both Base and StatsCollector ctors) not considered an error?

**Comment by plroebuck** [comment]
Why is 'statsCollector' class not 'StatsCollector'? [Class naming convention](http://snowdream.github.io/javascript-style-guide/javascript-style-guide/en/naming-conventions.html).

**Comment by plroebuck** [comment]
"lib/stats-collector.js#L29":
  Realize your code just migrated existing, but why is `start` event not using `runner.once`?

"lib/stats-collector.js#L58":
Would also be nice to move the `pending` handler prior to the `end` handler.
(i.e., let `start` and `end` be bookend events).

**Comment by craigtaub** [comment]
@plroebuck thanks for the help here

- regarding "test/reporters/list.spec.js", it required a specific event callback handler to run. Now that I have introduced a new callback handler on the same event it fails if the wrong handler is triggered. Its really not nice but I could not think of a better way to target it without touching live code. Will have another think if there is a cleaner way.
- your right will look at adding a throw for "missing runner argument" error.
- with naming followed conventions of camelCase as its a function (and is called as such), not a constructor or class. Let me know if you have any thoughts on this.
- I will investigate `start` using `runner.once` and let you know
- Completely agree with bookending, will change that

**Comment by craigtaub** [comment]
@plroebuck have made some updates based on your comments, please have a look. Thanks

File: lib/stats-collector.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    stats.start = new Date();
+  });
+
+  runner.on('suite', function(suite) {
+    stats.suites = stats.suites || 0;
+    suite.root || stats.suites++;
+  });
+
+  runner.on('test end', function() {
+    stats.tests = stats.tests || 0;
```
Comment (plroebuck) [inline lib/stats-collector.js]: So why exactly is this line needed?

File: lib/stats-collector.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    stats.tests++;
+  });
+
+  runner.on('pass', function(test) {
+    stats.passes = stats.passes || 0;
```
Comment (plroebuck) [inline lib/stats-collector.js]: So why exactly is this line needed?

File: lib/stats-collector.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    stats.passes++;
+  });
+
+  runner.on('fail', function(test, err) {
+    stats.failures = stats.failures || 0;
```
Comment (plroebuck) [inline lib/stats-collector.js]: So why exactly is this line needed?

File: lib/stats-collector.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    stats.failures++;
+  });
+
+  runner.on('pending', function() {
```
Comment (plroebuck) [inline lib/stats-collector.js]: Notice it doesn't exist for pending.

File: lib/stats-collector.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
+'use strict';
+
+/**
+ * Stats Collector
+ *
+ * Provides stats such as test duration,
+ * number of tests passed / failed etc.
+ *
+ * @public
+ * @api public
```
Comment (plroebuck) [inline lib/stats-collector.js]: Delete `@api` line - just use standard JSDoc - this was referenced by one of TJ's old packages.

File: lib/stats-collector.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
+'use strict';
+
```
Comment (plroebuck) [inline lib/stats-collector.js]:
```js
/**
 * Reporter statistics.
 *
 * @typedef {Object} StatsCollector
 * @property {number} suites - integer count of suites run.
 * @property {number} tests - integer count of tests run.
 * @property {number} passes - integer count of passing tests.
 * @property {number} pending - integer count of pending tests.
 * @property {number} failures - integer count of failed tests.
 */
```

File: lib/stats-collector.js [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+ * Stats Collector
+ *
+ * Provides stats such as test duration,
+ * number of tests passed / failed etc.
+ *
+ * @public
+ * @api public
+ * @memberof Mocha
+ * @param {Runner} runner
+ * @return {stats}
```
Comment (plroebuck) [inline lib/stats-collector.js]: `@return {StatsCollector}`

File: lib/stats-collector.js [hunk 8]
```diff
@@ -0,0 +1,65 @@
+'use strict';
+
+/**
+ * Stats Collector
```
Comment (plroebuck) [inline lib/stats-collector.js]: Initializes test statistics collector object.

File: lib/stats-collector.js [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    stats.tests = stats.tests || 0;
+    stats.tests++;
+  });
+
+  runner.on('pass', function(test) {
+    stats.passes = stats.passes || 0;
+    stats.passes++;
+  });
+
+  runner.on('fail', function(test, err) {
```
Comment (plroebuck) [inline lib/stats-collector.js]: Maybe just since no need for args?
`runner.on('fail', function() {`

File: lib/stats-collector.js [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+    stats.suites = stats.suites || 0;
+    suite.root || stats.suites++;
+  });
+
+  runner.on('test end', function() {
+    stats.tests = stats.tests || 0;
+    stats.tests++;
+  });
+
+  runner.on('pass', function(test) {
```
Comment (plroebuck) [inline lib/stats-collector.js]: Maybe just this since no need for args?
`runner.on('pass', function() {`

File: test/reporters/list.spec.js [hunk 11]
```diff
@@ -138,10 +138,15 @@ describe('List reporter', function() {
       var actual = {a: 'actual'};
       var expected = {a: 'expected'};
       var checked = false;
+      var stringifyDiffs = 'stringifyDiffObjs';
       var err;
       test = {};
       runner.on = runner.once = function(event, callback) {
```
Comment (plroebuck) [inline test/reporters/list.spec.js]: If this is set to both `.on` and `.once`, aren't we calling it twice for same issue?

File: test/reporters/list.spec.js [hunk 12]
```diff
@@ -138,10 +138,15 @@ describe('List reporter', function() {
...
-        if (!checked && event === 'fail') {
+        if (
+          callback.toString().includes(stringifyDiffs) && // handle correct callback
```
Comment (plroebuck) [inline test/reporters/list.spec.js]: Sorry. Still hate this. Are we sure this is the **only** reporter affected? No other way to make this work correctly?

**Comment by plroebuck** [comment]
Modify your PR  above as follows:
* Replace existing contents of the "Applicable Issues" section with the text below (so the issue will be closed automatically upon merge and label this as a minor enhancement).
```text
Fixes #3256
(semver-minor)
```

File: lib/stats-collector.js [hunk 13]
```diff
@@ -0,0 +1,65 @@
+'use strict';
+
+/**
+ * Stats Collector
+ *
+ * Provides stats such as test duration,
+ * number of tests passed / failed etc.
+ *
+ * @public
```
Comment (plroebuck) [inline lib/stats-collector.js]: Is this being exported somewhere? Otherwise, should be `@private`

File: lib/stats-collector.js [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+  runner.once('start', function() {
+    stats.start = new Date();
+  });
+
+  runner.on('suite', function(suite) {
+    stats.suites = stats.suites || 0;
+    suite.root || stats.suites++;
+  });
+
+  runner.on('test end', function() {
```
Comment (plroebuck) [inline lib/stats-collector.js]: Move to before `runner.once('end', ` such that events listed in order of occurrence.

Comment (plroebuck) [reply lib/stats-collector.js]: @boneskull, why does Mocha count "suites" as soon as they start, yet count "tests"  only when they complete?

File: lib/stats-collector.js (hunk 13 above)
Comment (craigtaub) [reply lib/stats-collector.js]: I assumed this was flagging the module/function itself as public (is that right?). Its exported and imported by base reporter.

File: lib/stats-collector.js (hunk 1 above)
Comment (craigtaub) [reply lib/stats-collector.js]: Yep not needed will remove them all.

File: test/reporters/list.spec.js (hunk 11 above)
Comment (craigtaub) [reply test/reporters/list.spec.js]: I dont think so. `once` touches `start`+`end` events and `on` touches everything else.
As the assertion here doesn't need `start` or `end` I believe I can remove the `once`.

File: lib/stats-collector.js (hunk 6 above)
Comment (craigtaub) [reply lib/stats-collector.js]: Added property docs below just below section on module.

File: lib/reporters/base.js [hunk 15]
```diff
@@ -11,6 +11,7 @@ var diff = require('diff');
 var ms = require('../ms');
 var utils = require('../utils');
 var supportsColor = process.browser ? null : require('supports-color');
+var statsCollector = require('../stats-collector');
```
Comment (plroebuck) [inline lib/reporters/base.js]: `var initStatsCollector = require('../stats-collector');`

File: test/reporters/list.spec.js [hunk 16]
```diff
@@ -138,10 +138,16 @@ describe('List reporter', function() {
       var actual = {a: 'actual'};
       var expected = {a: 'expected'};
       var checked = false;
+      var stringifyDiffs = 'stringifyDiffObjs';
       var err;
       test = {};
-      runner.on = runner.once = function(event, callback) {
-        if (!checked && event === 'fail') {
+      runner = createMockRunner('fail', 'fail', null, null, test);
```
Comment (plroebuck) [inline test/reporters/list.spec.js]: Maybe append `runner = null;` to the `afterEach()` function.

File: lib/reporters/base.js (hunk 15 above)
Comment (plroebuck) [inline lib/reporters/base.js]: `var createStatsCollector = require('../stats-collector');`

File: lib/reporters/base.js [hunk 17]
```diff
@@ -283,68 +282,31 @@ exports.list = function(failures) {
...
-    pending: 0,
-    failures: 0
-  });
   var failures = (this.failures = []);
 
   if (!runner) {
-    return;
+    throw new Error('Missing runner argument');
   }
+  this.stats = statsCollector(runner);
```
Comment (plroebuck) [inline lib/reporters/base.js]: `  this.stats = createStatsCollector(runner);`

File: lib/stats-collector.js [hunk 18]
```diff
@@ -0,0 +1,72 @@
+'use strict';
+
+/**
+ * Initializes test statistics collector object.
```
Comment (plroebuck) [inline lib/stats-collector.js]: * Creates test statistics collector object.

File: lib/stats-collector.js [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+ * @property {number} suites - integer count of suites run.
+ * @property {number} tests - integer count of tests run.
+ * @property {number} passes - integer count of passing tests.
+ * @property {number} pending - integer count of pending tests.
+ * @property {number} failures - integer count of failed tests.
+ * @property {Date} start - time when testing began.
+ * @property {Date} end - time when testing concluded.
+ * @property {number} duration - number of msecs that testing took.
+ */
+function initStatsCollector(runner) {
```
Comment (plroebuck) [inline lib/stats-collector.js]: `function createStatsCollector(runner) {`

Sorry for the naming waffling here...

File: lib/stats-collector.js [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+
+  runner.once('end', function() {
+    stats.end = new Date();
+    stats.duration = stats.end - stats.start;
+  });
+
+  return stats;
+}
+
+module.exports = initStatsCollector;
```
Comment (plroebuck) [inline lib/stats-collector.js]: `module.exports = createStatsCollector;`

File: test/reporters/list.spec.js [hunk 21]
```diff
@@ -138,10 +138,16 @@ describe('List reporter', function() {
...
+      runner.on = function(event, callback) {
+        if (
+          callback.toString().includes(stringifyDiffs) && // target correct fail event callback
```
Comment (plroebuck) [inline test/reporters/list.spec.js]: What is actual intent here? Are you lazily trying to see if `stringifyDiffObjs` _is_ the callback, or if the callback invokes `stringifyDiffObjs`?

File: lib/stats-collector.js (hunk 6 above)
Comment (plroebuck) [reply lib/stats-collector.js]: One last thing here. I gave the object description as "Reporter statistics.", but since we refer to it as "Test statistics" everywhere else, this should probably be changed to match.

File: test/reporters/list.spec.js [hunk 22]
```diff
@@ -138,10 +138,16 @@ describe('List reporter', function() {
       var actual = {a: 'actual'};
       var expected = {a: 'expected'};
       var checked = false;
+      var stringifyDiffs = 'stringifyDiffObjs';
```
Comment (plroebuck) [inline test/reporters/list.spec.js]: As this is used nowhere else, migrate it inside the event handler callback.

File: test/reporters/list.spec.js (hunk 21 above)
Comment (plroebuck) [reply test/reporters/list.spec.js]: Might also migrate this check to the end of the `if` condition. It's more expensive than the other two checks.

Comment (plroebuck) [inline test/reporters/list.spec.js]: Might also migrate this to the end of the `if` condition. It's much more expensive than the other two conditionals.

File: lib/stats-collector.js (hunk 13 above)
Comment (craigtaub) [reply lib/stats-collector.js]: Sorry I find this super confusing, the function itself is exported though (i.e. `module.exports = createStatsCollector`). Its publicly available for anything else to import from anywhere.
Do you have an example perhaps of something similar which is private?

File: test/reporters/list.spec.js (hunk 21 above)
Comment (craigtaub) [reply test/reporters/list.spec.js]: Yep will do

File: test/reporters/list.spec.js (hunk 22 above)
Comment (craigtaub) [reply test/reporters/list.spec.js]: But then the same value is declared and assigned "every single time" a callback runs (and with this test it happens quite a lot). Isn't it better its assigned beforehand and then re-used?

File: test/reporters/list.spec.js (hunk 21 above)
Comment (craigtaub) [reply test/reporters/list.spec.js]: The latter...callback invokes `stringifyDiffObjs`. Will update var name.

File: lib/stats-collector.js (hunk 6 above)
Comment (craigtaub) [reply lib/stats-collector.js]: Ok have now separated and updated them.

**Comment by craigtaub** [comment]
@plroebuck made some updates from your comments

File: test/reporters/list.spec.js [hunk 23]
```diff
@@ -138,15 +139,15 @@ describe('List reporter', function() {
...
       var err;
       test = {};
       runner = createMockRunner('fail', 'fail', null, null, test);
       runner.on = function(event, callback) {
         if (
-          callback.toString().includes(stringifyDiffs) && // target correct fail event callback
           !checked &&
-          event === 'fail'
+          event === 'fail' &&
+          callback.toString().includes(callbackWhichStringifies) // target correct fail event callback
```
Comment (plroebuck) [inline test/reporters/list.spec.js]: How about just ditch the variable altogether?
```js
          callback.toString().includes('stringifyDiffObjs') // target correct fail event callback
```

File: lib/stats-collector.js (hunk 13 above)
Comment (plroebuck) [reply lib/stats-collector.js]: Mea culpa. I was essentially documenting `@api private` rather than `@public`. Sorry for the confusion.

File: lib/stats-collector.js [hunk 24]
```diff
@@ -1,19 +1,21 @@
...
  * Provides stats such as test duration,
  * number of tests passed / failed etc.
  *
  * @public
  * @memberof Mocha
  * @param {Runner} runner
- * @return {StatsCollector}
- *
- * Reporter statistics.
+ * @return {stats}
```
Comment (plroebuck) [inline lib/stats-collector.js]: Revert this change.

File: lib/stats-collector.js [hunk 25]
```diff
@@ -1,19 +1,21 @@
...
+ */
+
+/**
+ * {stats} object properties
  *
- * @typedef {Object} StatsCollector
+ * @typedef {Object} stats
```
Comment (plroebuck) [inline lib/stats-collector.js]: Revert this change. We're documenting what we're creating. As the function is named `createStatsCollector()`, the `@typedef` should match - hence:
`@typedef {Object} StatsCollector`.

File: lib/stats-collector.js [hunk 26]
```diff
@@ -1,19 +1,21 @@
...
  * @memberof Mocha
  * @param {Runner} runner
- * @return {StatsCollector}
- *
- * Reporter statistics.
+ * @return {stats}
+ */
+
+/**
+ * {stats} object properties
```
Comment (plroebuck) [inline lib/stats-collector.js]:
```text
 * Test statistics
```

File: lib/stats-collector.js (hunk 25 above)
Comment (plroebuck) [reply lib/stats-collector.js]: `@typedef` docblock should **preceed** function docblock.

Invert them so function docblock is paired with the function.

**Comment by plroebuck** [comment]
Before merge, we should update the [third-party reporter documentation](https://github.com/mochajs/mocha/wiki/Third-party-reporters) page as follows:
1. Remove the `passes` and `failures` variables.
1. Change the `end` event handler to get pass/fail information from new `stats` object.
1. Remove `process.exit` call (which should not have been there to begin with).

**Comment by Bamieh** [comment]
@plroebuck @craigtaub can you ping me when the code is ready 🙏 ?

**Comment by plroebuck** [comment]
@craigtaub, have you had a chance to look at the requested changes yet?

**Comment by craigtaub** [comment]
@Bamieh yep no problem.

@plroebuck unfortunately not yet (just started new job), will let you know once I have (hoping to within few days).

**Comment by plroebuck** [comment]
After going through something else, I no longer think the stats collector should belong to the reporter. Rather it belongs to the runner (though `Base` reporter can keep a reference for convenience).

Could be placed in the `Runner` [constructor](https://github.com/mochajs/mocha/blob/d401ee198bd5026bc61641ad26eab31e6746713f/lib/runner.js#L66-L84), somewhere after the existing event handlers (or at the end of function).

Alternatively, it could be placed in `Mocha.prototype.run()` immediately after runner instantiation [here](https://github.com/mochajs/mocha/blob/d401ee198bd5026bc61641ad26eab31e6746713f/lib/mocha.js#L581), but the former seems more appropriate.

File: lib/stats-collector.js (hunk 26 above)
Comment (craigtaub) [reply lib/stats-collector.js]: @plroebuck  sorry I don't follow this comment, it is not a `Test` right?

File: test/reporters/list.spec.js (hunk 23 above)
Comment (craigtaub) [reply test/reporters/list.spec.js]: Sure will do.

File: test/reporters/list.spec.js [hunk 27]
```diff
@@ -32,6 +32,7 @@ describe('List reporter', function() {
   });
```
[the remark this answers is not in the document]
Comment (craigtaub) [reply test/reporters/list.spec.js]: Done

File: lib/stats-collector.js (hunk 25 above)
Comment (craigtaub) [reply lib/stats-collector.js]: @plroebuck I have now reverted this change.

**Comment by craigtaub** [comment]
@plroebuck regarding the Runner, it's a fair point. I did initially think about reporter vs runner but decided for simplicity sake (and I could not think of a better reason) to keep it in reporter.

How strongly do you feel that change needs to come in this PR?

We have abstracted the stats which feels like a good first step. I am happy to create a new issue referencing your comment which would move it to the Runner.

File: lib/stats-collector.js (hunk 26 above)
Comment (plroebuck) [reply lib/stats-collector.js]: [Like this](https://github.com/mochajs/mocha/blob/fe623c745c44085ee79b1978e98969454d662848/lib/stats-collector.js#L4):
```js
/**
 * Test statistics collector.
 *
 * @typedef {Object} StatsCollector
...
```

Comment (craigtaub) [reply lib/stats-collector.js]: My query is around the term "Test". What is this module testing?

**Comment by plroebuck** [comment]
> Regarding the Runner, it's a fair point. I did initially think about reporter vs runner but decided for simplicity sake (and I could not think of a better reason) to keep it in reporter.

Had done the same.

> How strongly do you feel that change needs to come in this PR?
>
> We have abstracted the stats which feels like a good first step. I am happy to create a new issue referencing your comment which would move the required functionality to the Runner. Personally I am a fan of splitting work into chunks and if you agree at least we could move this PR on.

Would really prefer to finish in this PR; thus far, test statistics collection was refactored into its own file, but not really moved from the `Base` reporter. Moving the stats collector to the `Runner` (and leaving a reference in `Base` reporter) is pretty trivial.

**Comment by craigtaub** [comment]
@plroebuck wasn't the notion to remove the connection from Base reporter only introduced 2 days ago? In't this why best practice is to aim for increments not big bang?

Honestly there is a real frustration to chasing PR approvals in such a way (for me feels like this PR has been open forever). Specially for volunteer work which has to be done in spare time. Changes in small chunks is surely ideal.

**Comment by plroebuck** [comment]
Moving it would appear to be less than 10 lines of code.
But if you're tired of it, I'll do it.

**Comment by craigtaub** [comment]
@plroebuck I have not tried it, that was not the point of my comment.
I won't have time for a few weeks but if you leave it I will get to it eventually. Feel free to make the changes yourself.

**Comment by craigtaub** [comment]
@plroebuck commits squashed.

File: lib/reporters/base.js [hunk 28]
```diff
@@ -283,68 +282,31 @@ exports.list = function(failures) {
...
-    pending: 0,
-    failures: 0
-  });
   var failures = (this.failures = []);
 
   if (!runner) {
-    return;
+    throw new Error('Missing runner argument');
   }
+  this.stats = getStats(); // attached so Reporters have a reference
```
Comment (plroebuck) [inline lib/reporters/base.js]: Why isn't this
`this.stats = runner.stats;`

Comment (craigtaub) [reply lib/reporters/base.js]: Yep good call, have made change + build passing. 
If theres nothing else could you approve it please and I will "Rebase and merge".

File: lib/reporters/base.js [hunk 29]
```diff
@@ -283,68 +281,31 @@ exports.list = function(failures) {
...
-    tests: 0,
-    passes: 0,
-    pending: 0,
-    failures: 0
-  });
   var failures = (this.failures = []);
 
   if (!runner) {
-    return;
+    throw new Error('Missing runner argument');
```
Comment (plroebuck) [inline lib/reporters/base.js]: Should _technically_ be `TypeError`

File: test/reporters/xunit.spec.js [hunk 30]
```diff
@@ -104,6 +105,7 @@ describe('XUnit reporter', function() {
           callback();
         }
       };
+      runner.stats = createStatsCollector(runner);
```
Comment (plroebuck) [inline test/reporters/xunit.spec.js]: Better here or L29.5?

File: test/reporters/xunit.spec.js [hunk 31]
```diff
@@ -339,6 +341,7 @@ describe('XUnit reporter', function() {
         }
       };
       runner.once = runner.on;
+      runner.stats = createStatsCollector(runner);
```
Comment (plroebuck) [inline test/reporters/xunit.spec.js]: Leave this to your call as partially outside your scope (don't remember "xunit.spec.js" being in previous commits).

The `runner` variable at L328 masks the one at L14. Would it be better to remove it and change L337-L344 to this?
```js
runner.on = runner.once = function(eventName, eventHandler) {
  // capture the event handler
  events[eventName] = eventHandler;
};
```

Comment (craigtaub) [reply test/reporters/xunit.spec.js]: Yes I agree its better to define `runner` (and attach stats) once and just re-use. I need to keep an eye out for that in these tests.
Have now made that change.

BTW yep the original commit didn't have this change, it came from the ownership changing from Reporter to Runner.

File: lib/stats-collector.js [hunk 32]
```diff
@@ -0,0 +1,73 @@
...
+  var stats = {
+    suites: 0,
+    tests: 0,
+    passes: 0,
+    pending: 0,
+    failures: 0
+  };
+
+  if (!runner) {
+    throw new Error('Missing runner argument');
```
Comment (plroebuck) [inline lib/stats-collector.js]: TypeError. Sorry, missed this one last go-round.

File: test/reporters/xunit.spec.js (hunk 31 above)
Comment (plroebuck) [reply test/reporters/xunit.spec.js]: > BTW yep the original commit didn't have this change, it came from the ownership changing from Reporter to Runner.

Any idea why? Both Base and Runner had a reference to the same "object". Who created it shouldn't have mattered, unless I'm really missing something.

File: lib/stats-collector.js (hunk 32 above)
Comment (craigtaub) [reply lib/stats-collector.js]: Ah ok, done.

File: test/reporters/xunit.spec.js (hunk 31 above)
Comment (craigtaub) [reply test/reporters/xunit.spec.js]: The Runner in this test is a mocked object. So the Reporter + Base require the mock to have a "stats" property they can attach to.
Does that make sense? I hope I haven't misunderstood what you meant.

Comment (craigtaub) [reply test/reporters/xunit.spec.js]: Hmm did you mean does it need the assignment? I think your right + no assignment is needed.

**Comment by craigtaub** [comment]
No worries @plroebuck . Apologies this one took so long. 
Hopefully my last set of changes made sense.
